Refactor runtime crates into feature-guarded modules

jcrossley3 opened the pull request:
To address #130

slinkydeveloper commented:
Ok the first thing I would do is to split this PR into several PRs, one per submodule :)

slinkydeveloper commented:
You also need to sign the DCO :)

slinkydeveloper commented on src/lib.rs:
```diff
@@ -40,6 +40,15 @@
 #![doc(html_root_url = "https://docs.rs/cloudevents-sdk/0.3.1")]
 #![deny(broken_intra_doc_links)]
 
+#[cfg(feature = "ce-actix")]
```
Just name the feature `actix` without prepending with ce? Same for the below modules

jcrossley3 replied:
That was my initial approach. As you can see from my [2nd commit](https://github.com/cloudevents/sdk-rust/pull/131/commits/14ac80ebfb09bd5cff527d0404ba3b5551a2029d), I had a conflict with `reqwest`. I'm happy to resolve that however you like, but I opted for a consistent prefix.

slinkydeveloper commented on src/actix/server_request.rs:
```diff
@@ -134,6 +135,17 @@ impl HttpRequestExt for HttpRequest {
     }
 }
 
+impl actix_web::FromRequest for Event {
+    type Config = ();
+    type Error = actix_web::Error;
+    type Future = Ready<std::result::Result<Event, actix_web::Error>>;
+
+    fn from_request(req: &HttpRequest, payload: &mut actix_web::dev::Payload) -> Self::Future {
+        // TODO: request_to_event(req, web::Payload(payload.take()))
+        ok(Event::default())
+    }
+}
+
```
Can we do this in another PR, after this PR to move the code?

jcrossley3 replied:
Sure thing. I can back it out.

jcrossley3 commented:
> You also need to sign the DCO :)

I signed all but the one I need help with. I can do that when I figure out how to fix it.

jcrossley3 commented:
> Ok the first thing I would do is to split this PR into several PRs, one per submodule. Also don't remove the old packages now, we can do that later.

So you prefer 1-PR-per-submodule to 1-commit-per-submodule as I currently have it? I can do that, but I'm curious why that's better? Once I back out the deletion of the existing packages and `FromRequest` impl, we'll only be left with the 4 submodule commits in this PR. Is that better than 4 PR's with one commit each?

jcrossley3 commented:
Ok, I just force-pushed changes that leave the old package dirs and don't add the `FromRequest` impl. Just one commit for each submodule feature. I left the `ce-` prefix in for now, and I tried to address the test failures as well.

slinkydeveloper commented on src/actix/headers.rs:
```diff
@@ -0,0 +1,68 @@
+use std::collections::HashMap;
+use std::str::FromStr;
+
+macro_rules! unwrap_optional_header {
```
Not for this PR, but maybe for future refactorings: a lot of these macros are copy paste, because the APIs are practically the same even if their names are different, so it would be nice to find a way to unify them

jcrossley3 replied:
Addressed in #146

slinkydeveloper commented on src/actix/mod.rs:
````diff
@@ -0,0 +1,57 @@
+//! }
+//! ```
+//!
+//! Check out the [cloudevents-sdk](https://docs.rs/cloudevents-sdk) docs for more details on how to use [`cloudevents::Event`]
````
This is not useful anymore

slinkydeveloper replied:
Same comment applies to other commits as well

slinkydeveloper commented on src/actix/mod.rs:
```diff
@@ -0,0 +1,57 @@
+//! This crate integrates the [cloudevents-sdk](https://docs.rs/cloudevents-sdk) with [Actix web](https://docs.rs/actix-web/) to easily send and receive CloudEvents.
```
Same comment applies to other commits as well

slinkydeveloper commented on src/actix/mod.rs:
```diff
@@ -0,0 +1,57 @@
+//!
+//! Check out the [cloudevents-sdk](https://docs.rs/cloudevents-sdk) docs for more details on how to use [`cloudevents::Event`]
+
+#![doc(html_root_url = "https://docs.rs/cloudevents-sdk/0.3.1")]
+#![deny(broken_intra_doc_links)]
```
I think these are not useful anymore

slinkydeveloper replied:
Same comment applies to other commits as well

slinkydeveloper reviewed (commented):
Can you try to run locally `cargo doc --open --all-features` to see if this breaks the doc build? You should also try `cargo publish --dry-run` to see if any publish check breaks (although I'm not sure you can do that without logging in)

slinkydeveloper commented:
Another thing I'm not 100% sure of is how this is gonna play out with our existing CI config, with our matrix of different targets... For example, the actix feature is not supposed to build with wasm

slinkydeveloper commented on cloudevents-sdk-actix-web/Cargo.toml:
```diff
@@ -13,11 +13,11 @@ categories = ["web-programming", "encoding", "web-programming::http-server"]
 
 [dependencies]
-cloudevents-sdk = { version = "0.3.0", path = ".." }
+cloudevents-sdk = { path = ".." }
```
You need to keep the version, it's required when publishing

jcrossley3 replied:
Does this matter if we're going to delete the workspaces? Or do you want to continue to publish all of them even though the logic is replicated in the root? The `cargo publish --dry-run` from the root worked fine.

slinkydeveloper replied:
> Or do you want to continue to publish all of them even though the logic is replicated in the root?

Just for another version, then we'll remove them

> https://github.com/vert-x3/vertx-eventbus-bridge-clients/tree/master/Go

you need to try `cargo publish --dry-run` from the subdirs as well

jcrossley3 replied:
Done. I didn't touch `warp` though, since it already lacked a `version`. Never published, I guess?

jcrossley3 commented:
> Can you try to run locally `cargo doc --open --all-features` to see if this breaks the doc build? You should also try `cargo publish --dry-run` to see if any publish check breaks (although I'm not sure you can do that without logging in)

Done. Fixed several broken links. I only tested the publish from the root and it seemed fine, abandoning only because of `--dry-run` and without requiring me to log in.

jcrossley3 commented:
> Another thing I'm not 100% sure of is how this is gonna play out with our existing CI config, with our matrix of different targets... For example, the actix feature is not supposed to build with wasm. I might be wrong, but it seems to me that the integrations are not tested anymore with this PR, without modifying the existing github actions config.

I added an `--all-features` to every step that had `--workspace`. This essentially runs all the tests twice. When we delete the workspace dirs, we can remove the `--workspace` option in the actions, too.

As far as the wasm32 build goes, I added the `cloudevents-reqwest` feature to the args, but I'm unable to see this working locally. I can't get past the following:

```console
[jim@localhost reqwest-wasm-example]$ npm run serve

> @ serve /home/jim/src/rust/sdk-rust/example-projects/reqwest-wasm-example
> webpack-dev-server

🧐  Checking for wasm-pack...

ℹ️  Installing wasm-pack 

ℹ ｢wds｣: Project is running at http://localhost:8080/
ℹ ｢wds｣: webpack output is served from /
ℹ ｢wds｣: Content not from webpack is served from /home/jim/src/rust/sdk-rust/example-projects/reqwest-wasm-example
✖ ｢wdm｣: Error: Rust compilation.
    at ChildProcess.<anonymous> (/home/jim/src/rust/sdk-rust/example-projects/reqwest-wasm-example/node_modules/@wasm-tool/wasm-pack-plugin/plugin.js:210:16)
    at ChildProcess.emit (events.js:376:20)
    at maybeClose (internal/child_process.js:1055:16)
    at Socket.<anonymous> (internal/child_process.js:441:11)
    at Socket.emit (events.js:376:20)
    at Pipe.<anonymous> (net.js:673:12)
```

slinkydeveloper commented:
> As far as the wasm32 build goes, I added the cloudevents-reqwest feature to the args, but I'm unable to see this working locally. I can't get past the following:

Mhhh that error is weird, perhaps the js deps needs to be updated

jcrossley3 commented:
> > As far as the wasm32 build goes, I added the cloudevents-reqwest feature to the args, but I'm unable to see this working locally. I can't get past the following:
> 
> Mhhh that error is weird, perhaps the js deps needs to be updated

Didn't help. I get the same error on `master`, so I don't think this PR broke it. Besides, all the checks are passing. ;)

I think I've addressed all your concerns. Let me know if you need anything else!
